---
title: Dragging 5 year old homologene to today
author: Ogan Mancarci
date: '2019-02-08'
slug: homologene-update
draft: true
bibliography: bibliography.bib
output:
  blogdown::html_page:
    toc: true
categories:
  - Programming
tags:
  - R
  - bioinformatics
---

## Homologene

I am the author of a small package called [homologene](https://github.com/oganm/homologene). The package wraps
the identically named [homologene database released by NCBI](https://www.ncbi.nlm.nih.gov/homologene). This database includes
information about genes that are homologues/orthologues of each other. It is structured
as a simple table. One can simply access it by doing

```{r,message = FALSE}
library(readr)
library(magrittr)
library(dplyr)

tmp = tempfile()
download.file(url = "ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.data", 
              destfile = tmp)

homologene = read_tsv(tmp,
                      col_names = c('HID',
                      			  'Taxonomy',
                      			  'Gene.ID',
                      			  'Gene.Symbol',
                      			  'Protein.GI',
                      			  'Protein.Accession'))

head(homologene)

```

In homologene package, a subset of this data is available for querying.

```{r}
head(homologene::homologeneData)
```

At the time I removed the `Protein.GI` and `Protein.Accession` since my work had no use
for them and limited the information to 7 popular model organisms to so that package
wouldn't be too bloated.

My use of the the package mostly consists of translating orthologues between mouse and human.

```{r}
homologene::human2mouse(c('ENO2','MOG','GZMH'))
```

In general I like using this method because it's fast, the syntax is simple and it's easy to extend it at a whim. A single set of gene symbols and IDs are used to identify genes and organism level data is tied to taxonomy identifiers which makes things easier. The obvious alternative to this would be to use `biomart` [@biomart] which to me, is less intuitive and clunky.

## What is wrong with homologene?

Not everyone is happy with using homologene for orthology detection. The database has not been updated since 2014. This means it is not created using the latest annotations. The reference genomes have been getting updates and gene symbols change fairly frequently. In fact we can see how many of the genes don't have the same official first name or ID by using the [geneSynonym](https://github.com/oganm/geneSynonym) package. This package includes 


```{r}
library(geneSynonym)
library(homologene)
library(purrr)

# get human gene IDs from the human synonym data shipped with geneSynonym
geneIDs = names(syno10090)
homologeneData %>%
    filter(Gene.ID %in% geneIDs) ->
    mouseGenes

# get first names from the human synonym data
syno10090[as.character(mouseGenes$Gene.ID)] %>%
    strsplit('\\|') %>% map_chr(1) ->
    firstNames

# set new names as a column
mouseGenes$newNames = firstNames

mistmatchedNewNames = 
    mouseGenes %>% filter(newNames != Gene.Symbol)

nrow(mistmatchedNewNames)

head(mistmatchedNewNames)
```

That is `r nrow(mistmatchedNewNames)` genes that have a different official symbol than they used to in 2014. And that is only in mouse. Considering we are looking at a total of `r nrow(mouseGenes)` genes here, this number isn't too much but it does mean that we will be missing a few genes if we were using gene symbols to for matching. Alas, since I wasn't wise enough 4 years ago when I started my projects, there are parts of my pipelines that still rely on gene symbol matching. 

Most common gene list I translate using homologene is my list of brain cell type markers that were created using mouse data [@mancarci2017cross].

```{r}
library(markerGeneProfile)
data("mouseMarkerGenesCombined")

mouseMarkerGenesCombined %>% 
    unlist %>%
    unique %>% 
    {. %in% mistmatchedNewNames$newNames} %>% 
    sum
```

```{r,include = FALSE}
mouseMarkerGenesCombined %>% 
    unlist %>%
    unique  %>% length
```


I will see that there are mistmatches. Again, this isn't much but it means that by default my pipelines are also effected by this issue.

## Updating homologene gene symbols




## Bibliography