---
title: Updating the good old Homologene database
author: Ogan Mancarci
date: '2019-02-10'
description: Updating the gene symbols and IDs of the homologene database
slug: homologene-update
bibliography: bibliography.bib
csl: nature.csl
output:
  blogdown::html_page:
    toc: true
categories:
  - Programming
tags:
  - R
  - bioinformatics
editor_options: 
  chunk_output_type: inline
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
```

## Homologene

I am the author of a small package called [homologene](https://github.com/oganm/homologene). The package wraps
the identically named [homologene database released by NCBI](https://www.ncbi.nlm.nih.gov/homologene). This database includes
information about genes that are homologues/orthologues of each other. It is structured
as a simple table. One can simply access it by doing

```{r,eval=FALSE}
library(readr)
library(magrittr)
library(dplyr)
dir.create('homologene-files',showWarnings = FALSE)
download.file(url = "ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.data", 
              destfile = 'homologene-files/homologene.data')

homologene = read_tsv('homologene-files/homologene.data',
                      col_names = c('HID',
                      			  'Taxonomy',
                      			  'Gene.ID',
                      			  'Gene.Symbol',
                      			  'Protein.GI',
                      			  'Protein.Accession'))

homologene

```

```{r,echo = FALSE,message=FALSE}
library(readr)
library(magrittr)
library(dplyr)
if(!file.exists('homologene-files/homologene.data')){
	download.file(url = "ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.data", 
              destfile = 'homologene-files/homologene.data')
	homologene = read_tsv('homologene-files/homologene.data',
                      col_names = c('HID',
                      			  'Taxonomy',
                      			  'Gene.ID',
                      			  'Gene.Symbol',
                      			  'Protein.GI',
                      			  'Protein.Accession'))
	saveRDS(homologene,'homologene-files/homologene.rds')
} else{
	homologene = readRDS('homologene-files/homologene.rds')
}


homologene
```

In the homologene package, a subset of this data is available for querying.

```{r}
head(homologene::homologeneData)
```

At the time I removed the `Protein.GI` and `Protein.Accession` since my work had no use
for them and limited the information to 7 popular model organisms to so that package
wouldn't be too bloated.

The package includes functions to translate orthologues between species with a one liner

```{r}
homologene::homologene(c('ENO2','MOG','GZMH'),
					   inTax = 9606, # human
					   outTax = 10090) # mouse
```

In general I like using this method because it's fast, the syntax is simple and it's easy to extend it at a whim. A single set of gene symbols and IDs are used to identify genes and organism level data is tied to taxonomy identifiers which makes things easier. The obvious alternative to this would be to use `biomart` [@biomart] which to me, is less intuitive and clunky.

## What is wrong with homologene?

Not everyone is happy with using homologene for orthology detection. The database has not been updated since 2014. This means it is not created using the latest annotations. The reference genomes have been getting updates and gene symbols change frequently. For example [here](https://github.com/oganm/geneSynonym/commit/c2b92a6965fd56d73c016cb0f31e5550b5be7da2) you can see a list of gene name changes that happened in a 15 day period at February 9th 2019.

That list above is generated by parsing [gene_info.gz](ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/) file provided by NCBI. I use this file to keep an up to date list of gene symbols and their synonyms. It comes in handy when I use data released by other researchers that only include gene symbols as identifiers. Using the data here, we can identify how many genes in homologene have changed their names, assuming the NCBI ids are the same. I will be using the data in the form it is made available in the [geneSyonym](https://github.com/oganm/geneSynonym) package.


```{r, message=FALSE}
library(geneSynonym)
library(homologene)
library(purrr)

# get mouse gene IDs from the mouse synonym data shipped with geneSynonym
gene_IDs = names(syno10090)

# get mouse genes from the homologene data
homologene::homologeneData %>%
    filter(Gene.ID %in% gene_IDs) ->
    mouse_genes

# get fisrt names (official symbol) from the mouse synonym data
syno10090[as.character(mouse_genes$Gene.ID)] %>%
    strsplit('\\|') %>%
	map_chr(1) ->
    first_names

# set new names as a column
mouse_genes$new_names = first_names

mistmatched_new_names = 
    mouse_genes %>% filter(new_names != Gene.Symbol)

# how many mismatches
nrow(mistmatched_new_names)

# take a peek at the first 10
head(mistmatched_new_names, 10)
```

That is `r nrow(mistmatched_new_names)` genes that have a different official symbol than they used to in 2014. And that is only in mouse. Considering this only corresponds to `r format(100*nrow(mistmatched_new_names)/nrow(mouse_genes),digits = 2)`% of all gene symbols, this number isn't too much but it does mean that we will be missing a few genes if we were using gene symbols for matching. Alas, since I wasn't wise enough 4 years ago when I started my projects, there are parts of my pipelines that still rely on gene symbol matching.

Most common transformation I make is translating my lists of brain cell type markers that were created using mouse data [@mancarci2017cross] to their human orthologues. I can see which markers are lost due to differences in gene symbols when making this transformation.

```{r}
# this package includes my marker gene list
# github.com/pavlidisLab/markerGeneProfile
library(markerGeneProfile)
data("mouseMarkerGenesCombined")

mouseMarkerGenesCombined %>% 
    unlist %>%
    unique %>% 
    {.[. %in% mistmatched_new_names$new_names]} 
```

```{r,include = FALSE}
mouseMarkerGenesCombined %>% 
    unlist %>%
    unique %>% 
    {. %in% mistmatched_new_names$new_names} %>% 
    sum -> mismatches

mouseMarkerGenesCombined %>% 
    unlist %>%
    unique  %>% length ->totalCount
```

That is `r mismatches` genes which corresponds to a `r format(mismatches/totalCount*100,digits = 2)`%. This is probably a more realistic image of real world implications since not all genes are equally important. Many genes that are prone to change are pseudogenes, non-coding transcripts etc. Such genes are less likely to come up in an analysis.

## Updating homologene gene IDs

One small thing we can do to improve this situation is to manually replace the old gene symbols with new ones. Assuming the gene IDs are the same, we could just replace the old names with the new. However, gene IDs are also prone to changes [@informationGeneFrequentlyAsked2018] and there is a whole different file that keeps track of these changes. We need to look into that file and update the gene IDs too.

```{r,eval = FALSE}
download.file(url = "ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_history.gz", 
              destfile = 'homologene-files/gene_history.gz')

gene_history = read_tsv('homologene-files/gene_history.gz',
						col_names = c('tax_id', 
									  'GeneID',
									  'Discontinued_GeneID',
									  'Discontinued_Symbol',
									  'Discontinue_Date'))

gene_history
```

```{r,echo = FALSE,message=FALSE}

if(!file.exists('homologene-files/gene_history.rds')){
	download.file(url = "ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_history.gz", 
              destfile = 'homologene-files/gene_history.gz')
	gene_history = read_tsv('homologene-files/gene_history.gz',
					   col_names = c('tax_id', 
					   			  'GeneID',
					   			  'Discontinued_GeneID',
					   			  'Discontinued_Symbol',
					   			  'Discontinue_Date'),skip = 1)
	saveRDS(gene_history,'homologene-files/gene_history.rds')
} else{
	gene_history = readRDS('homologene-files/gene_history.rds')
}

gene_history

```
One unverified assumption I am making is that once an ID is discontinued, it won't be used for another gene. Let's make sure that is actually the case before doing anything else




```{r}
# combine all the modern IDs from the geneSynonym package (same species as the homologene package)
modern_IDs  = list(syno10090,
				   syno10116,
				   syno6239,
				   syno7227,
				   syno7955,
				   syno9544,
				   syno9606) %>%
	lapply(names) %>%
	do.call(c,.)

gene_history %>% filter(Discontinued_GeneID %in% modern_IDs)
```

No current ID was ever discontinued. That's a good sign. I can just trace
the history of discontinued IDs and find the current IDs. Note that when I first tried
this here were 20 IDs here. 20 IDs that were discontinued since I last updated the 
`geneSynonym` package. Since all of these files area updated nightly it is important
to update the databases at the same time.

```{r}
discontinued_ids = homologeneData %>% 
	filter(Gene.ID %in% gene_history$Discontinued_GeneID)

dim(discontinued_ids)

unchanged_ids = homologeneData %>%  
	filter(!Gene.ID %in% gene_history$Discontinued_GeneID)


```

It's good to see that only a relatively small number of IDs were discontinued. We
can trace their history to get their current IDs (if it exists). To make this process
faster, we will also take a relevant subset of the `gene_history` frame.


```{r}

# get the earliest date where we see one of our ID change events from homologene
# since homologene was updated in 2014, this should be no earlier that that date
earlierst_date = gene_history %>%
	filter(Discontinued_GeneID %in% homologeneData$Gene.ID) %$%
	Discontinue_Date %>% 
	min

earlierst_date

```

This is interesting. Why does homologene includes gene IDs that were discontinued 4 years before it's creation? How many such IDs are out there

```{r}
# get all discontinuation events earlier than 2014. Events are coded as integers
# in YYYYMMDD format
gene_history %>%
	filter(Discontinue_Date < 20140000 & Discontinued_GeneID %in% homologeneData$Gene.ID)
```

Ok it's not so bad. Not sure what's the reason behind this but it doesn't seem like a catastrophic failure.

So now we can filter the `gene_history` to only include species that we are interested
in and events that happened after the earliest discontinuation date.

```{r}
relevant_gene_history = gene_history %>% filter(Discontinue_Date >= earlierst_date & 
													tax_id %in% homologene::taxData$tax_id)

dim(relevant_gene_history)
```

That's a much smaller search space. Now we can just trace the history of every single
gene with a different ID than before. I will probably end up including a version of this function to the
`geneSynonym` package as pretty much all gene lists are at least a little out of date.

```{r}
traceID = function(id){
	print(id)
	event = relevant_gene_history %>% filter(Discontinued_GeneID == id)
	if(nrow(event)>1){
		# just in case. if the same ID is discontinued twice, there is a problem...
		return("multiple events")
	}
	while(TRUE){
		# see if the new ID is discontinued as well
		next_event = relevant_gene_history %>%
			filter(Discontinued_GeneID == event$GeneID)
		if(nrow(next_event)==0){
			# if not, previous ID is the right one
			return(event$GeneID)
		} else if(nrow(next_event)>1){
			# just in case, if the same ID is discontinued twice, there is a problem...
			return("multiple events")
		} else if(nrow(next_event) == 1){
			# if the new IDs is discontinued, continue the loop and check if it has a parent
			event = next_event
		}
	}
}
```

```{r, eval = FALSE}
discontinued_ids$Gene.ID %>%
	sapply(traceID) ->
	new_ids

```


```{r,echo = FALSE}

if(!file.exists('homologene-files/new_ids.rds')){
	discontinued_ids$Gene.ID %>%
		sapply(traceID) ->
		new_ids
	saveRDS(new_ids, "homologene-files/new_ids.rds")
} else{
	new_ids = readRDS("homologene-files/new_ids.rds")
}
```

Not the most efficient code probably but it'll be run as a CRON job in the office machine at around 6 am on Sundays
so I don't really care that much about efficiency.

Lets do a few sanity checks

```{r}
# are there any ids that had multiple events
any(new_ids == "multiple events")

# are there any new ids that the geneSynonym package doesn't know about?
all(new_ids[new_ids != '-'] %in% modern_IDs)

# how many of the discontinued IDs end up having new IDs instead of getting removed
length(new_ids[new_ids != '-'])

# how many IDs are lost forever
length(new_ids[new_ids == '-'])

```

Good.

## Updating homologene gene symbols

Since we have the new ids now, we can start building our new `homologeneData2` frame
that will be added to the `homologene` package. I don't want to overwrite the original data
because people might use it expecting a perfect match with the NCBI database. Also if
I messed up here I don't want people to suffer without explicitly choosing to trust me 
instead of the actual database.


```{r}
# create a frame with new ids
discontinued_fix = data.frame(HID = discontinued_ids$HID,
							  Gene.Symbol = discontinued_ids$Gene.Symbol,
							  Taxonomy = discontinued_ids$Taxonomy,
							  Gene.ID = new_ids,
							  stringsAsFactors = FALSE)

# remove symbols that are discontinued
discontinued_fix %<>% filter(Gene.ID != '-')

head(discontinued_fix)

```

Before going further let's make sure we aren't doing anything crazy by peeking
at the [NCBI website](https://www.ncbi.nlm.nih.gov/gene/562072) for the id of Itbp1 gene we see there.

All looks fine, so we can put everything together and replace the gene symbols as well

```{r}
# recombine the two lists sort by homology group
homologeneData2 = 
	rbind(discontinued_fix,unchanged_ids) %>% 
	arrange(HID)

# change the names with the new names
modern_symbols = list(syno10090,
					  syno10116,
					  syno6239,
					  syno7227,
					  syno7955,
					  syno9544,
					  syno9606) %>% 
	lapply(function(x){
		strsplit(x,split = "\\|") %>% map_chr(1)
	}) %>% do.call(c,.)

modern_frame = tibble(modern_IDs,
					  modern_symbols)

new_symbols = 
	modern_frame$modern_symbols[match(homologeneData2$Gene.ID, modern_frame$modern_IDs)]

# a sanity check. all genes with different symbols
sum(homologeneData2$Gene.Symbol != new_symbols)

# another sanity check, differences in mouse. 
# we did this in the beginningas well
# note that the number is higher than before 
# since the new IDs allow more genes to be changed
sum((homologeneData2$Gene.Symbol != new_symbols)[homologeneData2$Taxonomy ==10090])

homologeneData2 %<>% 
	mutate(Gene.Symbol = modern_frame$modern_symbols[match(Gene.ID,modern_frame$modern_IDs)])

head(homologeneData2)

```
```{r, echo = FALSE}
saveRDS(homologeneData2,file = "homologene-files/homologeneData2.rds")
```

The new homologene dataset and associated functions to use them should be on Github in a few days.
Then I will set up a CRON job to update it weekly. The CRAN version will have to be perpetually out of date which is somewhat problematic, but it can't be helped. I may include a function that builds the current version and let users specify the version they want to use.
The new syntax will probably look like this:

```r
# translate using the old database
homologene::homologene(c('ENO2','MOG','GZMH'),
					   inTax = 9606,
					   outTax = 10090) 


# translate using the new database
homologene::homologene(c('ENO2','MOG','GZMH'),
					   inTax = 9606,
					   outTax = 10090,
					   db = homologene2) 

# translate using the latest database
homologene::update_homologene(path)
homologene::homologene(c('ENO2','MOG','GZMH'),
					   inTax = 9606,
					   outTax = 10090,
					   db = path)

```


Before the next CRAN update I want to include a wrapper for biomaRt that uses the same syntax as homologene to make queries in ensembl so that'll likely be the next post.



## Bibliography