---
title: Dragging 5 year old homologene to today
author: Ogan Mancarci
date: '2019-02-08'
slug: homologene-update
draft: true
bibliography: bibliography.bib
output:
  blogdown::html_page:
    toc: true
categories:
  - Programming
tags:
  - R
  - bioinformatics
---


<div id="TOC">
<ul>
<li><a href="#homologene">Homologene</a></li>
<li><a href="#what-is-wrong-with-homologene">What is wrong with homologene?</a></li>
<li><a href="#updating-homologene-gene-symbols">Updating homologene gene symbols</a></li>
<li><a href="#bibliography">Bibliography</a></li>
</ul>
</div>

<div id="homologene" class="section level2">
<h2>Homologene</h2>
<p>I am the author of a small package called <a href="https://github.com/oganm/homologene">homologene</a>. The package wraps
the identically named <a href="https://www.ncbi.nlm.nih.gov/homologene">homologene database released by NCBI</a>. This database includes
information about genes that are homologues/orthologues of each other. It is structured
as a simple table. One can simply access it by doing</p>
<pre class="r"><code>library(readr)
library(magrittr)
library(dplyr)

tmp = tempfile()
download.file(url = &quot;ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.data&quot;, 
              destfile = tmp)

homologene = read_tsv(tmp,
                      col_names = c(&#39;HID&#39;,
                                  &#39;Taxonomy&#39;,
                                  &#39;Gene.ID&#39;,
                                  &#39;Gene.Symbol&#39;,
                                  &#39;Protein.GI&#39;,
                                  &#39;Protein.Accession&#39;))

head(homologene)</code></pre>
<pre><code>## # A tibble: 6 x 6
##     HID Taxonomy Gene.ID Gene.Symbol Protein.GI Protein.Accession
##   &lt;int&gt;    &lt;int&gt;   &lt;int&gt; &lt;chr&gt;            &lt;int&gt; &lt;chr&gt;            
## 1     3     9606      34 ACADM          4557231 NP_000007.1      
## 2     3     9598  469356 ACADM        160961497 NP_001104286.1   
## 3     3     9544  705168 ACADM        109008502 XP_001101274.1   
## 4     3     9615  490207 ACADM        545503811 XP_005622188.1   
## 5     3     9913  505968 ACADM        115497690 NP_001068703.1   
## 6     3    10090   11364 Acadm          6680618 NP_031408.1</code></pre>
<p>In homologene package, a subset of this data is available for querying.</p>
<pre class="r"><code>head(homologene::homologeneData)</code></pre>
<pre><code>##   HID Gene.Symbol Taxonomy Gene.ID
## 1   3      acdh-8     6239  173979
## 2   3      acdh-7     6239  181758
## 3   5     acdh-12     6239  174180
## 4   6       kat-1     6239  174161
## 5   6     T02G5.4     6239  187992
## 6  13     R04B3.2     6239  180550</code></pre>
<p>At the time I removed the <code>Protein.GI</code> and <code>Protein.Accession</code> since my work had no use
for them and limited the information to 7 popular model organisms to so that package
wouldn’t be too bloated.</p>
<p>My use of the the package mostly consists of translating orthologues between mouse and human.</p>
<pre class="r"><code>homologene::human2mouse(c(&#39;ENO2&#39;,&#39;MOG&#39;,&#39;GZMH&#39;))</code></pre>
<pre><code>##   humanGene mouseGene humanID mouseID
## 1      ENO2      Eno2    2026   13807
## 2       MOG       Mog    4340   17441
## 3      GZMH      Gzmd    2999   14941
## 4      GZMH      Gzme    2999   14942
## 5      GZMH      Gzmg    2999   14944
## 6      GZMH      Gzmf    2999   14943</code></pre>
<p>In general I like using this method because it’s fast, the syntax is simple and it’s easy to extend it at a whim. A single set of gene symbols and IDs are used to identify genes and organism level data is tied to taxonomy identifiers which makes things easier. The obvious alternative to this would be to use <code>biomart</code> <span class="citation">(Durinck et al. 2005)</span> which to me, is less intuitive and clunky.</p>
</div>
<div id="what-is-wrong-with-homologene" class="section level2">
<h2>What is wrong with homologene?</h2>
<p>Not everyone is happy with using homologene for orthology detection. The database has not been updated since 2014. This means it is not created using the latest annotations. The reference genomes have been getting updates and gene symbols change fairly frequently. In fact we can see how many of the genes don’t have the same official first name or ID by using the <a href="https://github.com/oganm/geneSynonym">geneSynonym</a> package. This package includes</p>
<pre class="r"><code>library(geneSynonym)
library(homologene)
library(purrr)</code></pre>
<pre><code>## 
## Attaching package: &#39;purrr&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:magrittr&#39;:
## 
##     set_names</code></pre>
<pre class="r"><code># get human gene IDs from the human synonym data shipped with geneSynonym
geneIDs = names(syno10090)
homologeneData %&gt;%
    filter(Gene.ID %in% geneIDs) -&gt;
    mouseGenes

# get first names from the human synonym data
syno10090[as.character(mouseGenes$Gene.ID)] %&gt;%
    strsplit(&#39;\\|&#39;) %&gt;% map_chr(1) -&gt;
    firstNames

# set new names as a column
mouseGenes$newNames = firstNames

mistmatchedNewNames = 
    mouseGenes %&gt;% filter(newNames != Gene.Symbol)

nrow(mistmatchedNewNames)</code></pre>
<pre><code>## [1] 1410</code></pre>
<pre class="r"><code>head(mistmatchedNewNames)</code></pre>
<pre><code>##    HID Gene.Symbol Taxonomy   Gene.ID newNames
## 1   93      Epb4.2    10090     13828    Epb42
## 2  479        Daf2    10090     13137    Cd55b
## 3  660      Gm3934    10090 100042625  Gstp-ps
## 4  664     Gucy1b3    10090     54195  Gucy1b1
## 5  984       Clca3    10090     23844    Clca1
## 6 1137        Gbas    10090     14467 Nipsnap2</code></pre>
<p>That is 1410 genes that have a different official symbol than they used to in 2014. And that is only in mouse. Considering we are looking at a total of 21077 genes here, this number isn’t too much but it does mean that we will be missing a few genes if we were using gene symbols to for matching. Alas, since I wasn’t wise enough 4 years ago when I started my projects, there are parts of my pipelines that still rely on gene symbol matching.</p>
<p>Most common gene list I translate using homologene is my list of brain cell type markers that were created using mouse data <span class="citation">(Mancarci et al. 2017)</span>.</p>
<pre class="r"><code>library(markerGeneProfile)
data(&quot;mouseMarkerGenesCombined&quot;)

mouseMarkerGenesCombined %&gt;% 
    unlist %&gt;%
    unique %&gt;% 
    {. %in% mistmatchedNewNames$newNames} %&gt;% 
    sum</code></pre>
<pre><code>## [1] 46</code></pre>
<p>I will see that there are mistmatches. Again, this isn’t much but it means that by default my pipelines are also effected by this issue.</p>
</div>
<div id="updating-homologene-gene-symbols" class="section level2">
<h2>Updating homologene gene symbols</h2>
</div>
<div id="bibliography" class="section level2 unnumbered">
<h2>Bibliography</h2>
<div id="refs" class="references">
<div id="ref-biomart">
<p>Durinck, Steffen, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma, and Wolfgang Huber. 2005. “BioMart and Bioconductor: A Powerful Link Between Biological Databases and Microarray Data Analysis.” <em>Bioinformatics</em> 21: 3439–40.</p>
</div>
<div id="ref-mancarci2017cross">
<p>Mancarci, B Ogan, Lilah Toker, Shreejoy J Tripathy, Brenna Li, Brad Rocco, Etienne Sibille, and Paul Pavlidis. 2017. “Cross-Laboratory Analysis of Brain Cell Type Transcriptomes with Applications to Interpretation of Bulk Tissue Data.” <em>eNeuro</em>. Society for Neuroscience, ENEURO–0212.</p>
</div>
</div>
</div>
